### PR TITLE
initial commit of LDAP analyzer

### PR DESCRIPTION
**Pull request by mblove17**
Initial commit for LDAP write operations (modify, add, delete, modifyDN).  Tested on Microsoft Active Directory against the above operations.

### CODE REVIEW

**Comment by 0xxon** [comment]
Thanks a lot for taking the time to write a LDAP analyzer. We just discussed how to proceed with your pull request. I will structure this in first outlining our approach to merging protocol analyzers, and then giving a bit of high-level feedback on your code.

Protocol analyzers are one of the central components of Bro and they are exposed to a great variety of untrusted data. New protocol analyzers have to be very stable, and performant, cover an approproate part of the protocol, and we have to be sure that they will not cause issues in already existing installations before we merge them. Because of this, the bar for protocol analyzers to be merged is much higher than the bar for other contributions to Bro.

Generally, new protocol analyzers have to see some wide-spread testing on real-world traffic, before they can be included in Bro. Note that this holds true for analyzers contributed by core team members, which are generally tested by numerous large sites before they are merged into Bro.

As a first step to getting this added to Bro, we would recommend first making your protocol analyzer a Bro package (https://github.com/bro/packages), which can be installed by everyone using bro-pkg. This allows users running Bro 2.5 to test your analyzer and give you feedback about problems, missing features, crashes, etc.

Now for feedback about the current state of the analyzer:

* Is there a reason why you include the KRB types in your analyzer? I might have missed that, but I did not see them beig actually used in ldap-protocol.pac. Generally speaking, if you need access to types from KRB, there is probably a better way to include them than doing a direct copy; e.g. similar to how you included ASN.1.

* The documentation of the events should be extended to look similar to the documentation generally used for Bro events; this will let our doc infrastructure autogenerate documentation

* Whitespacing is not consistent, at least in main.bro you sometimes use spaces instead of tabs.

* Currently your analyzer does not include tests; analyzers should be accompagnied by test-cases raising and checking the results of all (or at least most) of the exposed events; for example of how to write test-cases see, e.g. testing/btest/scripts/base/protocols/ssl

In conclusion, we think the first step for this is to fix the small issues I addressed above, and to convert this into a package to easily enable toe community to test it. Does this sound fine to you?

**Comment by 0xxon** [comment]
I am going to close this for the moment - please feel free to re-open when you have done more work on this, or when you want to discuss what to do next.

**Comment by mblove17** [comment]
No problem, we're working on it now.  We'll give you a heads up when its
ready.  Thanks!

**Comment by juno0812** [comment]
@0xxon, is there an example Bro package that is a protocol analyzer?  I have only been able to see examples and instructions for Bro packages that are scripts.  Thanks!

**Comment by 0xxon** [comment]
Sorry - I forgot to answer to this. Looking at https://github.com/SoftwareConsultingEmporium/ldap-analyzer it seems like you figured it out :)

Thanks for committing this. Also - feel free to announce this on our mailing list to get people to run it.
